### DISCUSSION TITLE
Enable CORS for the public registry endpoints

### DISCUSSION THREAD

**Discussion by arcanis** [npm/feedback#117] [Registry] [answered]
Ref: https://github.com/npm/npm-registry-couchapp/issues/108#issuecomment-174095040; Raising again since I suspect the registry should now already have solid protections against DDoS that it didn't have in 2016.

The registry endpoints don't have CORS headers, requiring to go through a proxy to fetch packages (metadata + tarballs) from a browser. In particular, the two following endpoints would benefit from that:

#### `GET /package-name`

It would allow remote websites (for instance a GitHub Page-hosted application) to reference arbitrary metadata from a package (for instance I could use that to print the latest available version for some of my packages, or the author list, or [...]).

#### `GET /package-name/-/package-name.tgz`

Similarly, this would allow to build package-related tools without requiring expensive proxies; one classic example would be to display the diff between two packages, similar to [this](https://diff.intrinsic.com/).

Additionally, the next Yarn version may support running within a browser, and having CORS enabled on the package archives themselves would be useful to this effect. It's not blocking, since whether this happen or not we can provide stub packages on our side, but it would be interesting to evaluate how far we could push this experiment.

**Comment by simonplend** [comment]
The public API endpoints which are available on `api.npmjs.org` - e.g. for retrieving package download counts - already send CORS response headers. If CORS headers were also added to `registry.npmjs.org` it would make it straightforward to combine data from both public APIs in client side applications.

As @arcanis has described, adding CORS headers to specific public registry endpoints would really open things up for the community to build awesome things on top of npm data.

**Reply by Jolg42** [reply]
About the downloads API:
https://github.com/npm/feedback/discussions/60#discussioncomment-328683

**Comment by Jolg42** [comment]
Looks like CORS headers are missing for the registry indeed https://registry.npmjs.org/prisma

One workaround would be to use npms https://api.npms.io/v2/package/prisma

**Reply by bobbui** [reply]
would really love this to be supported

**Reply by Jolg42** [reply]
Maybe a better workaround for now is to have a proxy:

I did a simple one with Cloudflare Workers at https://npm.rapide.workers.dev/prisma
It's a simple proxy to the registry (example https://registry.npmjs.org/prisma)
The code is here https://gist.github.com/Jolg42/6ecfadf6beee644a207220085317dbfe

@simonplend also mentioned a simple and beautiful solution using Netlify's proxy feature https://docs.netlify.com/routing/redirects/rewrites-proxies/#proxy-to-another-service

**Reply by jasonkuhrt** [reply]
Another option: https://npmjs.cf/

**Reply by yanbutan** [reply]
> Maybe a better workaround for now is to have a proxy:
> 
> I did a simple one with Cloudflare Workers at https://npm.rapide.workers.dev/prisma It's a simple proxy to the registry (example https://registry.npmjs.org/prisma) The code is here https://gist.github.com/Jolg42/6ecfadf6beee644a207220085317dbfe
> 
> @simonplend also mentioned a simple and beautiful solution using Netlify's proxy feature https://docs.netlify.com/routing/redirects/rewrites-proxies/#proxy-to-another-service

The proxy https://registry.npmjs.org did not work for me but https://npmjs.cf/ did.

**Comment by MylesBorins** [comment]
Hey All,

I've opened a ticket with our engineering team to discuss this feature request and will get back to y'all

**Reply by Jolg42** [reply]
@jasonkuhrt will be curious to know if something is planned here 😉

**Reply by jasonkuhrt** [reply]
Yes! :)

**Reply by MylesBorins** [reply]
Hey all, we are continuing to discuss this internally but it is a bit more complicated than just enabling CORS. We need to think through what is our "public API surface", "what and how we want to document that API", and "which APIs enable CORS".

I've made a note in the future section of the roadmap around `next steps for APIs` to serve as a bookmark to not forget this effort. As the internal conversations get a bit clearer we'll scope a proper item on the roadmap to track the effort we'll be doing around this

**Reply by jakearchibald** [reply]
> Hey all, we are continuing to discuss this internally but it is a bit more complicated than just enabling CORS.

It doesn't need to be 😄 

Right now, cli and platform apps can use https://registry.npmjs.org/ just as the `npm` cli can, but the lack of CORS means web apps cannot.

Improving/changing the API seems like a good thing, but it's separate to enabling CORS.

**Reply by MylesBorins** [reply]
Thanks for the prod @jakearchibald . As we talk about Q2 planning and priorities I'll bring this up

**Reply by arcanis** [reply]
Any news? 🙂

**Reply by MylesBorins** [reply]
@arcanis this is still on our radar but has not been something we have been able to prioritize with our current efforts 100% focused on bringing webauthn to the market as quickly as possible. We do still have an open feature request and plan to do a review of all open feature requests in the near future as we open up a bit of capacity for other work.

**Reply by jakearchibald** [reply]
Any news?

**Reply by MylesBorins** [reply]
Good timing. We've finished going through security review and are planning the public roll out. We plan to enable CORS for getting the packument and the tarball.

We will likely plan a dark ship (temporary roll out) next week. Will update with timing so you can test.

**Reply by RDIL** [reply]
This broke extremely valuable websites like https://npm.anvaka.com ...

**Reply by arcanis** [reply]
Anvaka seems to have broken because https://npmjs.cf (mentioned in other subthreads) adds the `*` pattern to the CORS header even if it's already there, making it invalid (notice the duplicate header, that Chrome interprets as a combination of both):

<img width="387" alt="image" src="https://user-images.githubusercontent.com/1037931/163212092-18121685-2b0b-460e-af00-e367e4535875.png">

I pinged @terinjokes on Twitter to let him know.

**Reply by terinjokes** [reply]
Oh, I didn't get a heads up for this. I'll fix that today.

**Reply by terinjokes** [reply]
Should be resolved now.

**Reply by MylesBorins** [reply]
FYI we only did a dark ship today to specifically find if anything broke. We are planning full roll out next week

**Reply by terinjokes** [reply]
I'm curious to know more to be able to determine if I'll be broken again (and what "dark ship" means in this context). Fill free to send me a DM Myles.

**Reply by broofa** [reply]
@MylesBorins:   It looks like not all resources are being served with CORS headers.  Is that expected?  Specifically, I've got a branch of npmgraph.js.org that I've switched to use `registry.npmjs.org` (I've historically used @terinjokes's `npmjs.cf` mirror).  Note the broken `mime-db` request at https://npmgraph-git-npmjshost-npmgraph.vercel.app/?q=request.

![CleanShot 2022-04-13 at 11 56 49@2x](https://user-images.githubusercontent.com/164050/163251021-7a317a86-7567-453a-933a-872987b822e5.png)

**Reply by MylesBorins** [reply]
shared with engineering team, thanks for letting us know

**Comment by jasonkuhrt** [comment]
This works for us for now https://npmjs.cf/

**Reply by yanbutan** [reply]
This should get more upvotes. The only one that worked for me so far. Thanks @jasonkuhrt !

**Comment by MylesBorins** [comment, marked as answer]
CORS is launched and major issues resolved 🎉

**Reply by broofa** [reply]
Looks good from npmgraph's end.  This is awesome, thank you!

**Comment by naruaway** [comment]
Hi, I finally made "npm" work in browser ([npm-in-browser](https://github.com/naruaway/npm-in-browser)) thanks to this CORS header :tada:

However, I found that CORS is not enabled for "HTTP 404 Not Found" responses and this is causing a mismatch with real npm CLI in local machines in terms of its observable behavior.

Do you think we can make sure that `registry.npmjs.org` always returns appropriate CORS headers even for non `200` responses?

I filed an issue in npm-in-browser, which explains more details and reproductions: https://github.com/naruaway/npm-in-browser/issues/1

**Reply by broofa** [reply]
Yeah, I see this as well on npmgraph: 
<img width="875" alt="image" src="https://github.com/npm/feedback/assets/164050/4cc4c0be-825e-4c86-b755-1b627dbc2d55">

**Reply by broofa** [reply]
@MylesBorins Is this on your radar?  Seems like this is pretty clearly a bug.
